### PR TITLE
feature/multi valued headers

### PR DESCRIPTION
**Pull request by baywet**
DO NOT MERGE BEFORE https://github.com/microsoft/kiota/pull/2045
related https://github.com/microsoft/kiota/issues/2032

### CODE REVIEW

File: components/abstractions/src/main/java/com/microsoft/kiota/RequestHeaders.java [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+    /**
+     * Adds a header to the current request.
+     * @param key the key of the header to add.
+     * @param value the value of the header to add.
+     */
+    public void add(@Nonnull final String key, @Nonnull final String value) {
+        Objects.requireNonNull(key);
+        Objects.requireNonNull(value);
+        final String normalizedKey = normalizeKey(key);
+        if(headers.containsKey(normalizedKey)) {
```
Comment (andrueastman) [inline components/abstractions/src/main/java/com/microsoft/kiota/RequestHeaders.java:30]:
```suggestion
        if(headers.containsKey(key)) { //containsKey implicitly normalizes the key
```
no need to normalize the keys twice

File: components/abstractions/src/main/java/com/microsoft/kiota/RequestHeaders.java [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+        Objects.requireNonNull(value);
+        return headers.put(normalizeKey(key), new HashSet<>(value));
+    }
+    @Override
+	@Nonnull
+    public Set<String> remove(@Nonnull final Object key) {
+        if (key instanceof String) {
+            return headers.remove(normalizeKey((String)key));
+        } else {
+            return Collections.emptySet();
```
[the remark this answers is not in the document]
Comment (baywet) [reply components/abstractions/src/main/java/com/microsoft/kiota/RequestHeaders.java:88]: This is what I was originally doing but sonarcloud was not happy about it.

File: components/abstractions/src/main/java/com/microsoft/kiota/RequestHeaders.java [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+	 * Removes a value from a header
+	 * @param key the key of the header to remove the value from
+	 * @param value the value to remove
+	 * @return true if the value was removed, false otherwise
+	 */
+    public boolean remove(@Nonnull final String key, @Nonnull final String value) { 
+		Objects.requireNonNull(key);
+		Objects.requireNonNull(value);
+		final String normalizedKey = normalizeKey(key);
+		if(headers.containsKey(normalizedKey)) {
```
[the remark this answers is not in the document]
Comment (baywet) [reply components/abstractions/src/main/java/com/microsoft/kiota/RequestHeaders.java:101]: We need the normalized key below to remove the entry, doing it this way (not calling our local override) saves one normalization

File: components/abstractions/src/main/java/com/microsoft/kiota/RequestHeaders.java (hunk 1 above)
Comment (baywet) [reply components/abstractions/src/main/java/com/microsoft/kiota/RequestHeaders.java:30]: We need the normalized key below as well

File: components/abstractions/src/main/java/com/microsoft/kiota/RequestHeaders.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (andrueastman) [reply components/abstractions/src/main/java/com/microsoft/kiota/RequestHeaders.java:101]: Right. Sorry I missed that.

File: components/abstractions/src/main/java/com/microsoft/kiota/RequestHeaders.java [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+    @Override
+    public boolean containsValue(@Nonnull final Object value) {
+		Objects.requireNonNull(value);
+        return headers.containsValue(value);
+    }
+    @Override
+	@Nonnull
+    public Set<String> get(@Nonnull final Object key) {
+		Objects.requireNonNull(key);
+        if (key instanceof String) {
```
[the remark this answers is not in the document]
Comment (baywet) [reply components/abstractions/src/main/java/com/microsoft/kiota/RequestHeaders.java:69]: done. GitHub copilot apparently doesn't "understand" the context on that aspect....

**Review by andrueastman — APPROVED** [review]

**Pull request merged**